// node-red-contrib-mobius-flow-thingsboard/dist/thingsboard-drivers/thingsboard-mqtt-client.js
"use strict";
/*

 Copyright (c) 2018, IAconnects Technology Limited
 All rights reserved.

 Redistribution and use in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation and/or
    other materials provided with the distribution.

 3. Neither the name of the copyright holder nor the names of its contributors may be
    used to endorse or promote products derived from this software without specific prior
    written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 */
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const thingsboard_mqtt_gateway_1 = require("./thingsboard-mqtt-gateway");
class ThingsboardMqttClient extends events_1.EventEmitter {
    constructor(siteURL, gatewayId, accessToken, allowSelfCert) {
        super();
        const config = {
            site: siteURL,
            gatewayId,
            accessToken,
            keepAlive: 30,
            reconnectPeriod: 5,
            slowReconnectPeriod: 12,
            allowSelfCert,
        };
        this.mqttGateway = new thingsboard_mqtt_gateway_1.IThingsboardMQTTGateway(config);
        this.mqttGateway.on('connecting', () => {
            this.emit('connecting');
        });
        this.mqttGateway.on('disconnected', () => {
            this.emit('connecting');
        });
        this.mqttGateway.on('connected', () => {
            this.emit('connected');
        });
        this.mqttGateway.on('attributes', (topic, message) => {
            this.emit('attributes', topic, message);
        });
    }
    connect() {
        return this.mqttGateway.connect();
    }
    disconnect() {
        return this.mqttGateway.disconnect(true);
    }
    updateDeviceClientAttributes(deviceName, attributes) {
        return new Promise((resolve, reject) => {
            if (!this.mqttGateway.isConnected) {
                reject(new Error('Not connected to Thingsboard'));
            }
            else {
                this.mqttGateway.publish('v1/gateway/attributes', JSON.stringify({ [deviceName]: attributes }), 1)
                    .then(() => {
                    resolve();
                })
                    .catch((err) => {
                    reject(err);
                });
            }
        });
    }
    updateDeviceTelemetry(deviceName, timestamp, telemetry) {
        return new Promise((resolve, reject) => {
            if (!this.mqttGateway.isConnected) {
                reject(new Error('Not connected to Thingsboard'));
            }
            else {
                this.mqttGateway.publish('v1/gateway/telemetry', JSON.stringify({
                    [deviceName]: [
                        {
                            ts: timestamp,
                            values: telemetry,
                        },
                    ],
                }), 1)
                    .then(() => {
                    resolve();
                })
                    .catch((err) => {
                    reject(err);
                });
            }
        });
    }
    registerDeviceAsConnected(deviceName) {
        this.mqttGateway.publish('v1/gateway/connect', JSON.stringify({ device: deviceName }), 1);
    }
    registerDeviceAsDisconnected(deviceName) {
        this.mqttGateway.publish('v1/gateway/disconnect', JSON.stringify({ device: deviceName }), 1);
    }
}
exports.ThingsboardMqttClient = ThingsboardMqttClient;
