// node-red-contrib-mobius-flow-thingsboard/dist/thingsboard-drivers/thingsboard-http-client.js
"use strict";
/*

 Copyright (c) 2018, IAconnects Technology Limited
 All rights reserved.

 Redistribution and use in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation and/or
    other materials provided with the distribution.

 3. Neither the name of the copyright holder nor the names of its contributors may be
    used to endorse or promote products derived from this software without specific prior
    written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 */
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable-next-line:no-var-requires
const Client = require('node-rest-client').Client;
class ThingsboardHttpClient {
    constructor(siteURL, userName, password) {
        this.siteURL = siteURL;
        this.userName = userName;
        this.password = password;
        this.token = '';
        this.refreshToken = '';
        this.httpClient = new Client();
        this.getHttpTokenFromTB();
        this.httpTokenRefreshTimer = setInterval(() => {
            this.getHttpTokenFromTB();
        }, 5 * 60 * 1000);
    }
    closeHttpConnection() {
        clearInterval(this.httpTokenRefreshTimer);
    }
    hasToken() {
        return this.token !== '';
    }
    getDeviceByName(deviceName) {
        return new Promise((resolve, reject) => {
            const payload = {
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                    'X-Authorization': `Bearer ${this.token}`,
                },
            };
            const url = `https://${this.siteURL}/api/tenant/devices?deviceName=${deviceName}`;
            this.httpClient.get(url, payload, (responseData, response) => {
                if (response.statusCode === 200) {
                    resolve(responseData);
                }
                else {
                    response.errorMessage = `Failed to get device: ${deviceName}`;
                    reject(response);
                }
            }).on('error', (err) => {
                reject(err);
            });
        });
    }
    getAssetByName(assetName) {
        return new Promise((resolve, reject) => {
            const payload = {
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                    'X-Authorization': `Bearer ${this.token}`,
                },
            };
            const url = `https://${this.siteURL}/api/tenant/assets?assetName=${assetName}`;
            this.httpClient.get(url, payload, (responseData, response) => {
                if (response.statusCode === 200) {
                    resolve(responseData);
                }
                else {
                    response.errorMessage = `Failed to get device: ${assetName}`;
                    reject(response);
                }
            }).on('error', (err) => {
                reject(err);
            });
        });
    }
    getDeviceCredentials(deviceId) {
        return new Promise((resolve, reject) => {
            const payload = {
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                    'X-Authorization': `Bearer ${this.token}`,
                },
            };
            const url = `https://${this.siteURL}/api/device/${deviceId}/credentials`;
            this.httpClient.get(url, payload, (responseData, response) => {
                if (response.statusCode === 200) {
                    resolve(responseData);
                }
                else {
                    response.errorMessage = `Failed to get access token for device: ${deviceId}`;
                    reject(response);
                }
            }).on('error', (err) => {
                reject(err);
            });
        });
    }
    addDevice(deviceName, mobiusObject) {
        return new Promise((resolve, reject) => {
            const deviceLabel = mobiusObject.resources[1].pv !== '' ? mobiusObject.resources[1].pv : mobiusObject.description;
            const payload = {
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                    'X-Authorization': `Bearer ${this.token}`,
                },
                data: {
                    name: deviceName,
                    type: mobiusObject.profileName,
                    label: deviceLabel,
                    additionalInfo: mobiusObject.resources[3].pv,
                },
            };
            const url = `https://${this.siteURL}/api/device`;
            this.httpClient.post(url, payload, (responseData, response) => {
                if (response.statusCode === 200 || response.statusCode === 400) {
                    resolve(responseData);
                }
                else {
                    response.errorMessage = `Failed to add device: ${deviceName}`;
                    reject(response);
                }
            }).on('error', (err) => {
                reject(err);
            });
        });
    }
    addAsset(assetName, mobiusObject) {
        return new Promise((resolve, reject) => {
            const assetLabel = mobiusObject.resources[1].pv !== '' ? mobiusObject.resources[1].pv : mobiusObject.description;
            const payload = {
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                    'X-Authorization': `Bearer ${this.token}`,
                },
                data: {
                    name: assetName,
                    type: mobiusObject.profileName,
                    label: assetLabel,
                    additionalInfo: mobiusObject.resources[3].pv,
                },
            };
            const url = `https://${this.siteURL}/api/asset`;
            this.httpClient.post(url, payload, (responseData, response) => {
                if (response.statusCode === 200 || response.statusCode === 400) {
                    resolve(responseData);
                }
                else {
                    response.errorMessage = `Failed to add asset: ${assetName}`;
                    reject(response);
                }
            }).on('error', (err) => {
                reject(err);
            });
        });
    }
    updateDeviceClientAttributes(deviceName, attributes) {
        return new Promise((resolve, reject) => {
            this.getDeviceByName(deviceName)
                .then((response) => {
                return this.getDeviceCredentials(response.id.id);
            })
                .then((response) => {
                const deviceAccessToken = response.credentialsId;
                const payload = {
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    data: attributes,
                };
                const url = `https://${this.siteURL}/api/v1/${deviceAccessToken}/attributes`;
                this.httpClient.post(url, payload, (httpResponseData, httpResponse) => {
                    if (httpResponse.statusCode === 200) {
                        resolve(httpResponseData);
                    }
                    else {
                        httpResponse.errorMessage = 'Failed to update device attributes';
                        reject(httpResponse);
                    }
                }).on('error', (err) => {
                    reject(err);
                });
            });
        });
    }
    updateDeviceTelemetry(deviceName, timestamp, telemetry) {
        return new Promise((resolve, reject) => {
            this.getDeviceByName(deviceName)
                .then((response) => {
                return this.getDeviceCredentials(response.id.id);
            })
                .then((response) => {
                const deviceAccessToken = response.credentialsId;
                const payload = {
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    data: { ts: timestamp, values: telemetry },
                };
                const url = `https://${this.siteURL}/api/v1/${deviceAccessToken}/telemetry`;
                this.httpClient.post(url, payload, (httpResponseData, httpResponse) => {
                    if (httpResponse.statusCode === 200) {
                        resolve(httpResponseData);
                    }
                    else {
                        httpResponse.errorMessage = 'Failed to update device telemetry';
                        reject(httpResponse);
                    }
                }).on('error', (err) => {
                    reject(err);
                });
            });
        });
    }
    updateAssetTelemetry(assetName, timestamp, telemetry) {
        return new Promise((resolve, reject) => {
            this.getAssetByName(assetName)
                .then((response) => {
                const assetId = response.id.id;
                const payload = {
                    headers: {
                        'Content-Type': 'application/json',
                        'Accept': 'application/json',
                        'X-Authorization': `Bearer ${this.token}`,
                    },
                    data: { ts: timestamp, values: telemetry },
                };
                const url = `https://${this.siteURL}/api/plugins/telemetry/ASSET/${assetId}/timeseries/SERVER_SCOPE/`;
                this.httpClient.post(url, payload, (httpResponseData, httpResponse) => {
                    if (httpResponse.statusCode === 200) {
                        resolve(httpResponseData);
                    }
                    else {
                        httpResponse.errorMessage = 'Failed to update asset telemetry';
                        reject(httpResponse);
                    }
                }).on('error', (err) => {
                    reject(err);
                });
            });
        });
    }
    getHttpTokenFromTB() {
        const payload = {
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
            data: {
                username: this.userName,
                password: this.password,
            },
        };
        const url = `https://${this.siteURL}/api/auth/login`;
        this.httpClient.post(url, payload, (responseData, response) => {
            if (response.statusCode === 200) {
                this.token = JSON.parse(responseData.toString()).token;
                this.refreshToken = JSON.parse(responseData.toString()).refreshToken;
            }
            else {
                this.token = '';
                this.refreshToken = '';
            }
        }).on('error', (err) => {
            this.token = '';
            this.refreshToken = '';
        });
    }
}
exports.ThingsboardHttpClient = ThingsboardHttpClient;
